### PR TITLE
[SPARK-18593][SQL] JDBCRDD returns incorrect results for filters on CHAR of PostgreSQL

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

In Apache Spark 1.6.x, JDBCRDD returns incorrect results for filters on CHAR of PostgreSQL. The root cause is that **PostgreSQL returns space padded string** as a result. So, the post processing filter **`Filter (a#0 = A)` is evaluated false**. Spark 2.0.0 removes the post filter because it is already handled inside the database by `PushedFilters: [EqualTo(a,A)]`. This PR backports the relevant commits to fix this problem in Spark 1.6.x. All credits should go to the original authors.

**PostgreSQL Table & Query**
```sql
postgres=# \d t_char
       Table "public.t_char"
 Column |     Type      | Modifiers 
--------+---------------+-----------
 a      | character(10) | 

postgres=# \d t_varchar
          Table "public.t_varchar"
 Column |         Type          | Modifiers 
--------+-----------------------+-----------
 a      | character varying(10) | 

postgres=# select * from t_char where a='A';  // it works
     a      
------------
 A         
(1 row)

postgres=# select * from t_char where a='A         ';
     a      
------------
 A         
(1 row)
```

**Spark 1.6.3 Result**
```scala
scala> val t_char = sqlContext.read.option("user", "postgres").option("password", "rootpass").jdbc("jdbc:postgresql://localhost:5432/postgres", "t_char", new java.util.Properties())
t_char: org.apache.spark.sql.DataFrame = [a: string]

scala> val t_varchar = sqlContext.read.option("user", "postgres").option("password", "rootpass").jdbc("jdbc:postgresql://localhost:5432/postgres", "t_varchar", new java.util.Properties())
t_varchar: org.apache.spark.sql.DataFrame = [a: string]

scala> t_char.show
+----------+
|         a|
+----------+
|A         |
|AA        |
|AAA       |
+----------+

scala> t_varchar.show
+---+
|  a|
+---+
|  A|
| AA|
|AAA|
+---+

scala> t_char.filter(t_char("a")==="A").show  // This is the wrong result
+---+
|  a|
+---+
+---+

scala> t_char.filter(t_char("a")==="A         ").show
+----------+
|         a|
+----------+
|A         |
+----------+

scala> t_varchar.filter(t_varchar("a")==="A").show
+---+
|  a|
+---+
|  A|
+---+

scala> t_char.filter(t_char("a")==="A").explain
== Physical Plan ==
Filter (a#0 = A)
+- Scan JDBCRelation(jdbc:postgresql://localhost:5432/postgres,t_char,[Lorg.apache.spark.Partition;@2f65c341,{user=postgres, password=rootpass})[a#0] PushedFilters: [EqualTo(a,A)]
```

## How was this patch tested?

N/A (This is a backport of the commits).

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @maropu , @viirya , @HyukjinKwon , @rxin .

This is a backport to resolve SPARK-18593 in `branch-1.6`.

Although this is a correctness issue, there is a workaround for this issue. Users can use TEXT or VARCHAR. In addition, I'm not sure we will have Apache Spark 1.6.4.

I'm not sure if `branch-1.6` can have these kind of commits to fix an issue.

Please let me know your opinion.

**Comment by maropu** [comment]
@dongjoon-hyun Thanks for letting me know! I'm not familiar with a backport policy in spark though, if we can, I think it'd be better to do because it seems there are still some of users use spark 1.6.x. By the way, do we need to remove the added spaces of PostgreSQL `CHAR` in `JdbcDialect` or something when loading these data into Spark? ISTM that this PostgreSQL-specific behaivor makes users confused when processing the loaded data in Spark.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -165,8 +165,64 @@ private[sql] object JDBCRDD extends Logging {
    * @return A Catalyst schema corresponding to columns in the given order.
    */
   private def pruneSchema(schema: StructType, columns: Array[String]): StructType = {
-    val fieldMap = Map(schema.fields map { x => x.metadata.getString("name") -> x }: _*)
-    new StructType(columns map { name => fieldMap(name) })
+    val fieldMap = Map(schema.fields.map(x => x.metadata.getString("name") -> x): _*)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Is this change related?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -241,53 +297,23 @@ private[sql] class JDBCRDD(
...
-  private val filterWhereClause: String = {
-    val filterStrings = filters map compileFilter filter (_ != null)
-    if (filterStrings.size > 0) {
-      val sb = new StringBuilder("WHERE ")
-      filterStrings.foreach(x => sb.append(x).append(" AND "))
-      sb.substring(0, sb.length - 5)
-    } else ""
-  }
+  private val filterWhereClause: String =
+    filters.map(JDBCRDD.compileFilter).flatten.mkString(" AND ")
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Thank you for cc'ing me @dongjoon-hyun. I believe we need to backport https://github.com/apache/spark/pull/13743/files too.

**Comment by HyukjinKwon** [comment]
I am fine with this in terms of the change I proposed in the PR before except that the comment above.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: This is good. This is a correctness fixing. It may be more needed to backport.

**Comment by viirya** [comment]
@dongjoon-hyun Thanks for cc'ing me. Combined with https://github.com/apache/spark/commit/b22b20db640e9fac20c5d54cc83964dc74393821, I think it is worth backporting this to branch-1.6.

**Comment by dongjoon-hyun** [comment]
Thank you, @maropu , @HyukjinKwon , @viirya . I'll update like that. 

@maropu . I think we don't need to remove that here at least for this issue.

We don't expect Apache Spark 1.6.4, but at least this PR presents a solution for community using Apache Spark 1.6.x.

**Comment by dongjoon-hyun** [comment]
Oops. I didn't replace `spark` with `sqlContext`. I'll fix soon.

**Comment by HyukjinKwon** [comment]
Actually @dongjoon-hyun, this is a rather a question. Should we include other JIRAs in the title which this PR includes to backport?

**Comment by dongjoon-hyun** [comment]
Only your commits will be in `branch-1.6` with the original titles.
Since this PR has nothing, this PR will disappear completely when the committer merge this. I think so.

**Comment by dongjoon-hyun** [comment]
Hmm. Ur, @HyukjinKwon , maybe, the Spark script might not work like that I mentioned.

Anyway, in that case, the committer should backport these 6 commits separately instead of this PR.

For this issue, I think we can talk about later (when it's needed). What I want is just having these 6 commits in `branch-1.6` if possible.

**Comment by HyukjinKwon** [comment]
( I was just curious as I just remember tracking down with the blame button. It is no strong opinion.)

**Comment by dongjoon-hyun** [comment]
Hi, @rxin , @gatorsmile , @hvanhovell , @srowen .

Could you give some opinion? What I hoped here is having these 6 commits in `branch-1.6`.

But, if backports are not proper for this issue by the Apache Spark policy, I will just mark SPARK-18593 as Resolved with a fix version 2.0.0 and close this PR.

I know that It's not a good time for you to see this. Sorry for asking this.

**Comment by rxin** [comment]
Do these 6 commits have any conflicts? If no I can just cherrypick them.

**Comment by dongjoon-hyun** [comment]
Oh, thank you for considering. There are a few trivial conflicts.
One is the following about `WholeStageCodegenExec` testcase which I omitted here.

https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala#L256-L266

**Comment by rxin** [comment]
If you don't mind, it'd be better to have separate pull requests so we can link them easily across branches.

**Comment by dongjoon-hyun** [comment]
Sure! I think the authors also agree with that if it's allowed.

Hi, @maropu , @HyukjinKwon , @viirya .
Could you proceed to backport yours to `branch-1.6`?

**Comment by rxin** [comment]
If you can tell me the ones that don't conflict, I can just cherry-pick them without pull requests.

You can also submit pull requests for those that conflict btw. Don't need the original authors to.

**Comment by gatorsmile** [comment]
uh, we are fixing another bug in this area. See the PR: https://github.com/apache/spark/pull/15662

**Comment by dongjoon-hyun** [comment]
That seems to be for master (2.2)?

**Comment by dongjoon-hyun** [comment]
I see, @rxin. I will split them and tell you. Also I will make separate PRs keeping the original author. 

I'll ping you a few hour later!

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .

All 6 commits are not cherry-pickable. Especially, the last 2 commits has inevitable conflicts due to some wide commits (about `import order` and `syntax` cleanups.)

But, I can find the following clean cherry-pick sequences for the first 4 commits.

```
$ git cherry-pick -x 28112657ea5919451291c21b4b8e1eb3db0ec8d4
$ git cherry-pick -x 0f6936b5f1c9b0be1c33b98ffb62a72ae0c3e2a8
$ git cherry-pick -x 7f443a6879fa33ca8adb682bd85df2d56fb5fcda
$ git cherry-pick -x 2aad2d372469aaf2773876cae98ef002fef03aa3
$ git cherry-pick -x 554d840a9ade79722c96972257435a05e2aa9d88
$ git cherry-pick -x 8c1b867cee816d0943184c7b485cd11e255d8130
$ git cherry-pick -x 5c2682b0c8fd2aeae2af1adb716ee0d5f8b85135
$ git cherry-pick -x ad5b7cfcca7a5feb83b9ed94b6e725c6d789579b
$ git cherry-pick -x 94f7a12b3c8e4a6ecd969893e562feb7ffba4c24
$ git diff HEAD~9 --stat
sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala | 102 +++++++++++++++++++++++++++++++++++---------------------
sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala                     |  68 ++++++++++++++++++++++++++++++++-----
sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala                  |  15 +++++++++
 3 files changed, 139 insertions(+), 46 deletions(-)
```

All are related commits in the above three files. If you don't mind, could you cherry-pick the above. The followings are the titles of them.

1. [SPARK-12236][SQL] JDBC filter tests all pass if filters are not really pushed down.
2. [SPARK-12249][SQL] JDBC non-equality comparison operator not pushed down.
3. [SPARK-12314][SQL] isnull operator not pushed down for JDBC datasource.
4. [SPARK-12315][SQL] isnotnull operator not pushed down for JDBC datasource.
5. Style fix for the previous 3 JDBC filter push down commits. (@rxin)
6. [SPARK-12446][SQL] Add unit tests for JDBCRDD internal functions 
7. [SPARK-12409][SPARK-12387][SPARK-12391][SQL] Support AND/OR/IN/LIKE push-down filters for JDBC
8. [SPARK-12409][SPARK-12387][SPARK-12391][SQL] Refactor filter pushdown for JDBCRDD and add few filters (Liang-Chi Hsieh)
9. [SPARK-10180][SQL] JDBC datasource are not processing EqualNullSafe filter 

After your cherry-picking, I will create two PRs for the remaining inevitable-conflict commits.

**Comment by dongjoon-hyun** [comment]
Or, may I create just 6 PRs sequentially?

**Comment by dongjoon-hyun** [comment]
Ping, @rxin .

**Comment by gatorsmile** [comment]
I can help you review the backported PRs. Please backport the PRs one by one and cc me.

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile. I will create like that!

**Comment by dongjoon-hyun** [comment]
Thank you all so much for giving your attentions and opinions here.

I asked the one which is not allowed here. Please refer #16127 for more details, too.

I'm closing this now.

Sorry for all your inconvenience.
